downloader/src/main/kotlin/Downloader.kt

```diff
@@ -130,8 +130,9 @@ object Downloader {
         val vcsMark = TimeSource.Monotonic.markNow()
 
         try {
-            // Cargo in general builds from source tarballs, so we prefer source artifacts to VCS.
-            if (pkg.id.type != "Cargo" || pkg.sourceArtifact == RemoteArtifact.EMPTY) {
+            // Cargo in general builds from source tarballs, so we prefer source artifacts over VCS, but still use VCS
+            // if no source artifact is given.
+            if (pkg.id.type != "Cargo" || (pkg.id.type == "Cargo" && pkg.sourceArtifact == RemoteArtifact.EMPTY)) {
```

Comment:
`pkg.id.type == "Cargo"` is obvious in the or branch, it doesn't clarify anything for me. If we preferred `a || (!a && b)` over `a || b` we would have to change a lot of if conditions.

Reply:
In the original code, it confused me that the `pkg.sourceArtifact == RemoteArtifact.EMPTY` condition is there at all, because in this branch we do not care about the source artifact, so why checking it? It only becomes meaningful if you also take the previous condition into account, which is what I wanted to emphasize in this specific case.

But I'll change it, maybe by introducing a variable with a speaking name.

Reply:
Ideally the downloader should not have to know about any special handling for Cargo packages, but that's a more complex change that we hopefully can look into next year.